File: bus_app.py
from flask import Flask, jsonify, render_template
from flask_cors import CORS
import requests
import os
from dotenv import load_dotenv
from datetime import datetime

app = Flask(__name__)
CORS(app)
load_dotenv()
api_key = os.getenv("API_KEY")

@app.route("/", methods=["GET"])
def home():
    return render_template("bus-predictions.html")


def get_data(api_key: str, base: str, **kw):
    """Return JSON data from given request"""
    base_url = f"http://mbus.ltp.umich.edu/bustime/api/v3/{base}?key={api_key}&format=json"
    for key, value in kw.items():
        base_url += f"&{key}={value}"
    return requests.get(base_url).json()

@app.route("/bus-prediction", methods=["GET"])
def get_predictions():
    """Return predictions for Northbound buses from CCTC as JSON"""
    base = "getpredictions"
    northbound_from_cctc = [["BB", "NORTHBOUND", "C250"], ["NW", "NORTHBOUND", "C251"], ["CN", "NORTHBOUND", "C250"]]
    stop_names = {
        "C250": "CCTC South",
        "C251": "CCTC North"    
    }
    

    result = []


    for route_id, direction, stop_id in northbound_from_cctc:
        prediction_data = get_data(api_key, base, rt=route_id, stpid=stop_id, tmres="s")
        
        if "prd" in prediction_data["bustime-response"]:
            for prediction in prediction_data["bustime-response"]["prd"]:
                if prediction["rtdir"] == direction:
                    result.append({
                        "stop_name": stop_names[stop_id],
                        "route_id": route_id,
                        "direction": direction,
                        "stop_id": stop_id,
                        "vehicle_id": prediction['vid'],
                        "arrival_time": prediction['prdctdn']
                    })
    
    return jsonify(result)

if __name__ == "__main__":
    app.run(debug=True)

File: bus_app.js
const express = require('express');
const cors = require('cors');
const axios = require('axios');
require('dotenv').config();

const app = express();
const port = process.env.PORT || 3000;
const apiKey = process.env.API_KEY;

app.use(cors());
app.use(express.static('public'));

// Serve the HTML file
app.get('/', (req, res) => {
    res.sendFile(__dirname + '/bus-predictions.html');
});

// Function to fetch data from the API
async function getData(apiKey, base, params = {}) {
    const baseUrl = `http://mbus.ltp.umich.edu/bustime/api/v3/${base}?key=${apiKey}&format=json`;
    const urlParams = new URLSearchParams(params).toString();
    const fullUrl = `${baseUrl}&${urlParams}`;

    try {
        const response = await axios.get(fullUrl);
        return response.data;
    } catch (error) {
        console.error('Error fetching data:', error);
        throw error;
    }
}

// Route to fetch bus predictions
app.get('/bus-prediction', async (req, res) => {
    const base = 'getpredictions';
    const northboundFromCctc = [
        ["BB", "NORTHBOUND", "C250"],
        ["NW", "NORTHBOUND", "C251"],
        ["CN", "NORTHBOUND", "C250"]
    ];

    const stopNames = {
        "C250": "CCTC South",
        "C251": "CCTC North"
    };

    const result = [];

    try {
        for (const [routeId, direction, stopId] of northboundFromCctc) {
            const predictionData = await getData(apiKey, base, { rt: routeId, stpid: stopId, tmres: "s" });

            if (predictionData["bustime-response"] && predictionData["bustime-response"]["prd"]) {
                for (const prediction of predictionData["bustime-response"]["prd"]) {
                    if (prediction.rtdir === direction) {
                        result.push({
                            stop_name: stopNames[stopId],
                            route_id: routeId,
                            direction: direction,
                            stop_id: stopId,
                            vehicle_id: prediction.vid,
                            arrival_time: prediction.prdctdn
                        });
                    }
                }
            }
        }

        res.json(result);
    } catch (error) {
        res.status(500).json({ error: 'Failed to fetch bus predictions' });
    }
});

app.listen(port, () => {
    console.log(`Server running on http://localhost:${port}`);
});
